src/util/createValidation.ts

```diff
@@ -137,24 +137,17 @@ export default function createValidation(config: {
 
     const shouldSkip = skipAbsent && isAbsent(value);
 
-    if (!options.sync) {
-      try {
-        Promise.resolve(!shouldSkip ? test.call(ctx, value, ctx) : true).then(
-          handleResult,
-          handleError,
-        );
-      } catch (err: any) {
-        handleError(err);
-      }
-
-      return;
-    }
-
     let result: ReturnType<TestFunction>;
     try {
-      result = !shouldSkip ? test.call(ctx, value, ctx) : true;
-
-      if (typeof (result as any)?.then === 'function') {
+      result = shouldSkip ? true : test.call(ctx, value, ctx);
+      const isPromise = typeof (result as any)?.then === 'function';
+      if (isPromise) {
+        if (!options.sync) {
+          return Promise.resolve(result).then(
+            handleResult,
+            handleError,
+          );
+        }
```

Comment:
```suggestion
        if (options.sync)
          throw Error(...)
           
        return result.then(
          handleResult,
          handleError,
        );
```

I think this is all you actually need here? the `sync` check now only serves to throw if there is a promise when there shouldn't be